File: cleanupPdfFilenames.js
const fs = require('fs').promises;
const path = require('path');

/**
 * Test function to rename a single PDF file
 * @param {string} originalFilename - Original filename to test
 * @returns {string} - New filename after processing
 */
function testRename(originalFilename) {
  console.log(`Testing rename for: ${originalFilename}`);
  
  // Find the 6th underscore
  let underscoreCount = 0;
  let cutIndex = -1;
  
  for (let i = 0; i < originalFilename.length; i++) {
    if (originalFilename[i] === '_') {
      underscoreCount++;
      if (underscoreCount === 6) {
        cutIndex = i;
        break;
      }
    }
  }
  
  if (cutIndex === -1) {
    console.log(`No 6th underscore found in: ${originalFilename}`);
    return originalFilename; // Return original if less than 6 underscores
  }
  
  // Extract everything before the 6th underscore and add .pdf extension
  const baseName = originalFilename.substring(0, cutIndex);
  const newFilename = baseName + '.pdf';
  
  console.log(`Original: ${originalFilename}`);
  console.log(`New:      ${newFilename}`);
  console.log(`Cut at position: ${cutIndex} (6th underscore)`);
  
  return newFilename;
}

/**
 * Generate new filename by removing text after 6th underscore
 * @param {string} filename - Original filename
 * @returns {string} - New filename
 */
function generateNewFilename(filename) {
  // Find the 6th underscore
  let underscoreCount = 0;
  let cutIndex = -1;
  
  for (let i = 0; i < filename.length; i++) {
    if (filename[i] === '_') {
      underscoreCount++;
      if (underscoreCount === 6) {
        cutIndex = i;
        break;
      }
    }
  }
  
  if (cutIndex === -1) {
    return filename; // Return original if less than 6 underscores
  }
  
  // Extract everything before the 6th underscore and add .pdf extension
  const baseName = filename.substring(0, cutIndex);
  return baseName + '.pdf';
}

/**
 * Recursively find all PDF files in a directory
 * @param {string} dirPath - Directory path to search
 * @returns {Array} - Array of file objects with full path and relative path
 */
async function findAllPdfFiles(dirPath) {
  const pdfFiles = [];
  
  async function searchDirectory(currentPath, relativePath = '') {
    try {
      const items = await fs.readdir(currentPath, { withFileTypes: true });
      
      for (const item of items) {
        const fullPath = path.join(currentPath, item.name);
        const relPath = path.join(relativePath, item.name);
        
        if (item.isDirectory()) {
          // Recursively search subdirectories
          await searchDirectory(fullPath, relPath);
        } else if (item.isFile() && item.name.toLowerCase().endsWith('.pdf')) {
          pdfFiles.push({
            fullPath: fullPath,
            relativePath: relPath,
            filename: item.name,
            directory: currentPath
          });
        }
      }
    } catch (error) {
      console.error(`Error reading directory ${currentPath}:`, error.message);
    }
  }
  
  await searchDirectory(dirPath);
  return pdfFiles;
}

/**
 * Rename all PDF files in the pdf_downloads folder
 * @param {boolean} dryRun - If true, only show what would be renamed without actually renaming
 */
async function renameAllPdfFiles(dryRun = true) {
  try {
    console.log('Starting PDF file renaming process...');
    console.log(`Mode: ${dryRun ? 'DRY RUN (no actual renaming)' : 'ACTUAL RENAMING'}`);
    
    const downloadsDir = path.join(__dirname, 'pdf_downloads');
    
    // Check if pdf_downloads directory exists
    try {
      await fs.access(downloadsDir);
    } catch (error) {
      console.error('pdf_downloads directory not found!');
      return;
    }
    
    console.log(`Searching for PDF files in: ${downloadsDir}`);
    
    // Find all PDF files
    const pdfFiles = await findAllPdfFiles(downloadsDir);
    console.log(`Found ${pdfFiles.length} PDF files to process`);
    
    if (pdfFiles.length === 0) {
      console.log('No PDF files found to rename');
      return;
    }
    
    let renamedCount = 0;
    let skippedCount = 0;
    let errorCount = 0;
    const results = [];
    
    for (const file of pdfFiles) {
      try {
        const newFilename = generateNewFilename(file.filename);
        
        if (newFilename === file.filename) {
          console.log(`⏭️  Skipping (no change needed): ${file.relativePath}`);
          skippedCount++;
          results.push({
            original: file.relativePath,
            new: file.relativePath,
            status: 'skipped',
            reason: 'No 6th underscore found or no change needed'
          });
          continue;
        }
        
        const newFullPath = path.join(file.directory, newFilename);
        
        // Check if target file already exists
        try {
          await fs.access(newFullPath);
          console.log(`⚠️  Warning: Target file already exists, skipping: ${newFilename}`);
          skippedCount++;
          results.push({
            original: file.relativePath,
            new: path.join(path.dirname(file.relativePath), newFilename),
            status: 'skipped',
            reason: 'Target file already exists'
          });
          continue;
        } catch (error) {
          // Target file doesn't exist, safe to rename
        }
        
        if (!dryRun) {
          // Actually rename the file
          await fs.rename(file.fullPath, newFullPath);
          console.log(`✅ Renamed: ${file.relativePath} → ${path.join(path.dirname(file.relativePath), newFilename)}`);
        } else {
          console.log(`🔍 Would rename: ${file.relativePath} → ${path.join(path.dirname(file.relativePath), newFilename)}`);
        }
        
        renamedCount++;
        results.push({
          original: file.relativePath,
          new: path.join(path.dirname(file.relativePath), newFilename),
          status: dryRun ? 'would_rename' : 'renamed',
          reason: 'Successfully processed'
        });
        
      } catch (error) {
        console.error(`❌ Error processing ${file.relativePath}:`, error.message);
        errorCount++;
        results.push({
          original: file.relativePath,
          new: null,
          status: 'error',
          reason: error.message
        });
      }
    }
    
    // Save results to file
    const resultsData = {
      summary: {
        total_files: pdfFiles.length,
        renamed_count: renamedCount,
        skipped_count: skippedCount,
        error_count: errorCount,
        dry_run: dryRun,
        processed_date: new Date().toISOString()
      },
      results: results
    };
    
    const resultsPath = path.join(__dirname, `pdf_rename_results_${dryRun ? 'dryrun' : 'actual'}_${new Date().toISOString().slice(0, 10)}.json`);
    await fs.writeFile(resultsPath, JSON.stringify(resultsData, null, 2), 'utf8');
    
    console.log('\n=== Summary ===');
    console.log(`Total files processed: ${pdfFiles.length}`);
    console.log(`${dryRun ? 'Would be renamed' : 'Renamed'}: ${renamedCount}`);
    console.log(`Skipped: ${skippedCount}`);
    console.log(`Errors: ${errorCount}`);
    console.log(`Results saved to: ${resultsPath}`);
    
    if (dryRun) {
      console.log('\n💡 This was a dry run. To actually rename files, call renameAllPdfFiles(false)');
    }
    
    return resultsData;
    
  } catch (error) {
    console.error('Error in renameAllPdfFiles:', error);
    throw error;
  }
}

// Run the script if this file is executed directly
if (require.main === module) {
  const args = process.argv.slice(2);
  
  if (args.includes('--test')) {
    // Test mode - test a sample filename
    console.log('=== Testing filename transformation ===\n');
    
    const testFilenames = [
      '국회본회의 회의록_047627_제20대_제354회_제13차_20171124_34.  에너지산업클러스터의 지정 및 육성에 관한 특별법안(장병완 의원 대표발의)(장병완ㆍ주승용ㆍ전혜숙ㆍ유동수ㆍ김경진ㆍ이동섭ㆍ박준영ㆍ정인화ㆍ박지원ㆍ김수민ㆍ김동철ㆍ박주선ㆍ송기석ㆍ이찬열ㆍ신용현ㆍ노웅래ㆍ이상돈ㆍ이태규ㆍ윤호중ㆍ천정배ㆍ윤영일ㆍ이개호ㆍ인재근ㆍ이채익ㆍ홍의락ㆍ최경환(국)ㆍ김민기ㆍ김중로ㆍ백재현ㆍ조배숙ㆍ권은희ㆍ손금주 의원 발의).pdf',
    ];
    
    for (const filename of testFilenames) {
      testRename(filename);
      console.log('---\n');
    }
  } else if (args.includes('--actual')) {
    // Actual renaming
    renameAllPdfFiles(false)
      .then(results => {
        console.log('\nPDF file renaming completed!');
      })
      .catch(error => {
        console.error('PDF renaming failed:', error);
        process.exit(1);
      });
  } else {
    // Default: dry run
    renameAllPdfFiles(true)
      .then(results => {
        console.log('\nDry run completed! Use --actual flag to perform actual renaming.');
      })
      .catch(error => {
        console.error('PDF renaming dry run failed:', error);
        process.exit(1);
      });
  }
}

module.exports = { 
  testRename, 
  generateNewFilename, 
  findAllPdfFiles, 
  renameAllPdfFiles 
};

File: cleanupPdfFilenames.py
import json
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Any, Tuple
import argparse

class PdfFilenameCleanup:
    def __init__(self):
        self.base_dir = Path(__file__).parent
        self.downloads_dir = self.base_dir / 'pdf_downloads'

    def test_rename(self, original_filename: str) -> str:
        """Test function to rename a single PDF file"""
        print(f'Testing rename for: {original_filename}')
        
        # Find the 6th underscore
        underscore_count = 0
        cut_index = -1
        
        for i, char in enumerate(original_filename):
            if char == '_':
                underscore_count += 1
                if underscore_count == 6:
                    cut_index = i
                    break
        
        if cut_index == -1:
            print(f'No 6th underscore found in: {original_filename}')
            return original_filename  # Return original if less than 6 underscores
        
        # Extract everything before the 6th underscore and add .pdf extension
        base_name = original_filename[:cut_index]
        new_filename = base_name + '.pdf'
        
        print(f'Original: {original_filename}')
        print(f'New:      {new_filename}')
        print(f'Cut at position: {cut_index} (6th underscore)')
        
        return new_filename

    def generate_new_filename(self, filename: str) -> str:
        """Generate new filename by removing text after 6th underscore"""
        # Find the 6th underscore
        underscore_count = 0
        cut_index = -1
        
        for i, char in enumerate(filename):
            if char == '_':
                underscore_count += 1
                if underscore_count == 6:
                    cut_index = i
                    break
        
        if cut_index == -1:
            return filename  # Return original if less than 6 underscores
        
        # Extract everything before the 6th underscore and add .pdf extension
        base_name = filename[:cut_index]
        return base_name + '.pdf'

    def find_all_pdf_files(self, dir_path: Path) -> List[Dict[str, Any]]:
        """Recursively find all PDF files in a directory"""
        pdf_files = []
        
        def search_directory(current_path: Path, relative_path: str = ''):
            try:
                for item in current_path.iterdir():
                    if item.is_dir():
                        # Recursively search subdirectories
                        rel_path = str(Path(relative_path) / item.name) if relative_path else item.name
                        search_directory(item, rel_path)
                    elif item.is_file() and item.suffix.lower() == '.pdf':
                        rel_path = str(Path(relative_path) / item.name) if relative_path else item.name
                        pdf_files.append({
                            'full_path': item,
                            'relative_path': rel_path,
                            'filename': item.name,
                            'directory': item.parent
                        })
            except Exception as error:
                print(f'Error reading directory {current_path}: {error}')
        
        search_directory(dir_path)
        return pdf_files

    def rename_all_pdf_files(self, dry_run: bool = True) -> Dict[str, Any]:
        """Rename all PDF files in the pdf_downloads folder"""
        try:
            print('Starting PDF file renaming process...')
            print(f'Mode: {"DRY RUN (no actual renaming)" if dry_run else "ACTUAL RENAMING"}')
            
            # Check if pdf_downloads directory exists
            if not self.downloads_dir.exists():
                print('pdf_downloads directory not found!')
                return {'error': 'pdf_downloads directory not found'}
            
            print(f'Searching for PDF files in: {self.downloads_dir}')
            
            # Find all PDF files
            pdf_files = self.find_all_pdf_files(self.downloads_dir)
            print(f'Found {len(pdf_files)} PDF files to process')
            
            if not pdf_files:
                print('No PDF files found to rename')
                return {'message': 'No PDF files found'}
            
            renamed_count = 0
            skipped_count = 0
            error_count = 0
            results = []
            
            for file_info in pdf_files:
                try:
                    new_filename = self.generate_new_filename(file_info['filename'])
                    
                    if new_filename == file_info['filename']:
                        print(f'⏭️  Skipping (no change needed): {file_info["relative_path"]}')
                        skipped_count += 1
                        results.append({
                            'original': file_info['relative_path'],
                            'new': file_info['relative_path'],
                            'status': 'skipped',
                            'reason': 'No 6th underscore found or no change needed'
                        })
                        continue
                    
                    new_full_path = file_info['directory'] / new_filename
                    
                    # Check if target file already exists
                    if new_full_path.exists():
                        print(f'⚠️  Warning: Target file already exists, skipping: {new_filename}')
                        skipped_count += 1
                        new_relative = str(Path(file_info['relative_path']).parent / new_filename)
                        results.append({
                            'original': file_info['relative_path'],
                            'new': new_relative,
                            'status': 'skipped',
                            'reason': 'Target file already exists'
                        })
                        continue
                    
                    if not dry_run:
                        # Actually rename the file
                        file_info['full_path'].rename(new_full_path)
                        new_relative = str(Path(file_info['relative_path']).parent / new_filename)
                        print(f'✅ Renamed: {file_info["relative_path"]} → {new_relative}')
                    else:
                        new_relative = str(Path(file_info['relative_path']).parent / new_filename)
                        print(f'🔍 Would rename: {file_info["relative_path"]} → {new_relative}')
                    
                    renamed_count += 1
                    results.append({
                        'original': file_info['relative_path'],
                        'new': new_relative,
                        'status': 'would_rename' if dry_run else 'renamed',
                        'reason': 'Successfully processed'
                    })
                    
                except Exception as error:
                    print(f'❌ Error processing {file_info["relative_path"]}: {error}')
                    error_count += 1
                    results.append({
                        'original': file_info['relative_path'],
                        'new': None,
                        'status': 'error',
                        'reason': str(error)
                    })
            
            # Save results to file
            results_data = {
                'summary': {
                    'total_files': len(pdf_files),
                    'renamed_count': renamed_count,
                    'skipped_count': skipped_count,
                    'error_count': error_count,
                    'dry_run': dry_run,
                    'processed_date': datetime.now().isoformat()
                },
                'results': results
            }
            
            date_str = datetime.now().strftime('%Y-%m-%d')
            mode_str = 'dryrun' if dry_run else 'actual'
            results_path = self.base_dir / f'pdf_rename_results_{mode_str}_{date_str}.json'
            
            with open(results_path, 'w', encoding='utf-8') as f:
                json.dump(results_data, f, ensure_ascii=False, indent=2)
            
            print('\n=== Summary ===')
            print(f'Total files processed: {len(pdf_files)}')
            print(f'{"Would be renamed" if dry_run else "Renamed"}: {renamed_count}')
            print(f'Skipped: {skipped_count}')
            print(f'Errors: {error_count}')
            print(f'Results saved to: {results_path}')
            
            if dry_run:
                print('\n💡 This was a dry run. To actually rename files, use --actual flag')
            
            return results_data
            
        except Exception as error:
            print(f'Error in rename_all_pdf_files: {error}')
            raise error

def main():
    """Main function with command line argument parsing"""
    parser = argparse.ArgumentParser(description='Clean up PDF filenames by removing text after 6th underscore')
    parser.add_argument('--test', action='store_true', 
                       help='Test mode - test sample filename transformations')
    parser.add_argument('--actual', action='store_true', 
                       help='Perform actual renaming (default is dry run)')
    
    args = parser.parse_args()
    
    cleanup = PdfFilenameCleanup()
    
    if args.test:
        # Test mode - test a sample filename
        print('=== Testing filename transformation ===\n')
        
        test_filenames = [
            '국회본회의 회의록_047627_제20대_제354회_제13차_20171124_34.  에너지산업클러스터의 지정 및 육성에 관한 특별법안(장병완 의원 대표발의)(장병완ㆍ주승용ㆍ전혜숙ㆍ유동수ㆍ김경진ㆍ이동섭ㆍ박준영ㆍ정인화ㆍ박지원ㆍ김수민ㆍ김동철ㆍ박주선ㆍ송기석ㆍ이찬열ㆍ신용현ㆍ노웅래ㆍ이상돈ㆍ이태규ㆍ윤호중ㆍ천정배ㆍ윤영일ㆍ이개호ㆍ인재근ㆍ이채익ㆍ홍의락ㆍ최경환(국)ㆍ김민기ㆍ김중로ㆍ백재현ㆍ조배숙ㆍ권은희ㆍ손금주 의원 발의).pdf'
        ]
        
        for filename in test_filenames:
            cleanup.test_rename(filename)
            print('---\n')
    
    else:
        # Perform renaming (dry run by default, actual if --actual flag is used)
        dry_run = not args.actual
        results = cleanup.rename_all_pdf_files(dry_run)
        
        if not args.actual:
            print('\nDry run completed! Use --actual flag to perform actual renaming.')

if __name__ == "__main__":
    main()